text/0024-cron.md

```diff
@@ -0,0 +1,196 @@
+
+Running `qri cron add` for the same dataset name `b5/dataset_name` will overwrite the previous job entry with new values. cron jobs establish their next execution date by examining the more recent of two time sources:
+1. last cron job execution
+2. dataset commit timestamp
+
+In this case a cron job has never run, so the scheduled date is derived from the commit timestamp (in this example the dataset was committed 4 hours ago). Any time either a new dataset version is created or the job is run, the next cron execution time is re-evaluated & scheduled. This way if a user manually runs `qri update me/dataset_name`, cron will reset to use that time as an interval basis, avoiding excessive updates.
+
+The flags for `qri cron add` are be a subset of those available for `qri update`. Anything `qri update` can do, cron should work the same way: 
+```
+$ qri update me/github_stars --config=org,qri-io,repo,qri-io --secrets=token,$GITHUB_TOKEN
+$ qri cron add me/github_stars --config=org,qri-io,repo,qri-io --secrets=token,$GITHUB_TOKEN
+```
+
+We can see scheduled jobs with `qri cron list`:
+```
+$ qri cron list
+1. b5/dataset_name
+   scheduled: every 8 hours
+   next: Tues. April 11, 11:30:00
+   4 hours from now
+```
+The conventions for `list` should follow all other "list command conventions" for pagination, but listing order should be chronological according to next scheduled update soonest update first, fathest update last.
+
+We can start the qri cron service like this:
+```
+$ qri cron start
+cron update service started. stop it with:
+$ qri cron stop
+```
+
+If the current operating system doesn't yet support a daemonized service, the deamon should start in-process and show a warning message explaining to run this command with some equivalent of `nohup`. We should work to make as few people as possible see that message. As shown in the dialog, the cron daemon must be stoppable with `qri cron stop`.
+
+With cron running Qri will run a lightweight background loop that checks for update events and executes them, logging the results to the event log. To see a log of cron events, run:
+```
+$ qri cron log
+* me/dataset_name
+  succeeded 3h4m ago
+* me/github_stats
+  failed 2d16h27m ago
+  message: error: network timeout
+ # ...
+```
+
+This will show a list of successful & failed job executions
+
+### Scheduling shell scripts
+
+Scheduling datasets directly is the ideal workflow in terms of portability, but a whole set of use cases open by adding the capacity to schedule & execute shell scripts within the same cron environment. While most of this is obviously available by directly using an operating system `cron` service, having scripts for dataset updates collected in one place makes scheduling & maintenance easier. For that reason, shell scripts should be supported as a first-class citizen:
```

Comment:
Alternative implementation: add support for "hooks", like git has. Then, I could choose to run a shell script only when `cron` has found new data.

Related: do shell scripts have a one-to-one relationship with datasets, or do we support a shell script updates multiple datasets?

Reply:
OooOOOh hooks is a very interesting concept to bring up while we're having this conversation.

Many use cases will create new versions as the _result_ of running a script, not just a consequence of a dataset update completing. Concrete examples with high-level steps for completion:

#### libp2p_node_count
* port forward our cloud-based prometheus server to my local machine
* run `qri update me/lip2p_node_count`
* close the ssh tunnel

#### edgi_webmon_sitemap
* spin up a copy of `walk` with a preconfigured job file that creates a sitemap
* wait a number of hours
* once the process exits it'll write a file called `sitemap.json`
* run `qri save --body sitemap.json me/some_sitemap`

In both cases we're using a shell script to generate data, and then adding it to Qri, so the script should come first.

That being said, the question of how hooks might intersect with this conversation should be addressed. We've also had requests for _webhooks_, both of which are basically "run this when x event happens".

To me the mental model here is "scheduling" or "orchestration". cron is pulling Qri's puppet strings, whereas hooks are reactionary, and should be build as a separate service that tails additions to the qri event log.

> do shell scripts have a one-to-one relationship with datasets, or do we support a shell script updates multiple datasets?

To me there is no correlation at all. I think it's totally fine to run a script that doesn't update a qri dataset at all, so it's a zero-or-more relationship. shell scripts can and should call `qri update` directly, and if they do so multiple times, that's fine.

In documentation I think we would _strongly_ recommend against running multiple updates in the same script, as it's no longer conceptually transactional. You'd have to roll back a partially-successful script by hand. Thankfully there's a full audit trail for this sort of thing.